### PR TITLE
Listen for PullRequestEvents

### PR DESCRIPTION
**Pull request by ryanslade**
Instead of PullRequestActivityEvents

Also, ensure we don't fire both Activity and Pull Request events apart from the move from Needs Work to Unapproved.

Part of: https://github.com/sourcegraph/sourcegraph/issues/9139

### CODE REVIEW

File: src/main/java/com/sourcegraph/webhook/EventSerializer.java [hunk 1]
```diff
@@ -112,12 +115,16 @@ public ApplicationEvent getEvent() {
     }
 
     private void buildApplicationEvent(ApplicationEvent event) {
-        payload.addProperty("createdDate", RFC3339.format(event.getDate()));
+        payload.addProperty("createdDate", event.getDate().toInstant().toEpochMilli());
```
Comment (ryanslade) [inline src/main/java/com/sourcegraph/webhook/EventSerializer.java:118]: This is safe as none of our go code was loading this value before

**Comment by ryanslade** [comment]
This is already deployed on our dev server

File: src/main/java/com/sourcegraph/webhook/WebhookListener.java [hunk 2]
```diff
@@ -35,8 +38,30 @@
...
+                return;
+            }
+            if (previousStatus != PullRequestParticipantStatus.NEEDS_WORK) {
+                return;
+            }
+            key =  "pr:participant:status";
+        }
+
+        if (event instanceof PullRequestActivityEvent) {
+            key = "pr:activity:event";
```
Comment (mrnugget) [inline src/main/java/com/sourcegraph/webhook/WebhookListener.java:63]: The old default key was `pr:activity:status` — is this change intentional?

File: src/main/java/com/sourcegraph/webhook/WebhookListener.java [hunk 3]
```diff
@@ -35,8 +38,30 @@
     }
 
     @EventListener
-    public void onPullRequestEvent(PullRequestActivityEvent event) {
-        String key = "pr:activity:status";
+    public void onPullRequestEvent(PullRequestEvent event) {
+        String key = "pr:event";
+
+        // When review status changes, two events are fired in most cases
```
Comment (mrnugget) [inline src/main/java/com/sourcegraph/webhook/WebhookListener.java:44]: What does "in most cases" mean? It's a bit unnerving to read "most" without further qualification in code documentation 😄

File: src/main/java/com/sourcegraph/webhook/WebhookListener.java [hunk 4]
```diff
@@ -35,8 +38,30 @@
...
+        if (event instanceof PullRequestParticipantStatusUpdatedEvent) {
+            PullRequestAction action = event.getAction();
+            PullRequestParticipantStatus previousStatus = ((PullRequestParticipantStatusUpdatedEvent) event).getPreviousStatus();
+            if (action != PullRequestAction.UNAPPROVED) {
+                return;
+            }
+            if (previousStatus != PullRequestParticipantStatus.NEEDS_WORK) {
+                return;
+            }
+            key =  "pr:participant:status";
```
Comment (mrnugget) [inline src/main/java/com/sourcegraph/webhook/WebhookListener.java:59]: Do we need a new key? Or can't we reuse the `pr:activity:reviewers` key that I see further down? I'd rather have use introduce some abstraction here, than to handle this special case here and in our codebase.

**Review by mrnugget — APPROVED** [review]
Left some comments/thoughts, no blockers though.

I don't know the class hierarchy (a `PullRequestActivityEvent` _is a_ `PullRequestEvent` I assume?) to confidently say that this is wrong or not. But I assume you tested it, so approving to unblock.

File: src/main/java/com/sourcegraph/webhook/WebhookListener.java (hunk 2 above)
Comment (ryanslade) [reply src/main/java/com/sourcegraph/webhook/WebhookListener.java:63]: Yes, that's just cleaning things up. Our current go code ignores it anyway but an incoming PR uses the new values.

**Comment by ryanslade** [comment]
> Left some comments/thoughts, no blockers though.
> 
> I don't know the class hierarchy (a `PullRequestActivityEvent ` _is a_ `PullRequestEvent` I assume?) to confidently say that this is wrong or not. But I assume you tested it, so approving to unblock.

Yes, `PullRequestActivityEvent` is a subclass of `PullRequestEvent`

File: src/main/java/com/sourcegraph/webhook/WebhookListener.java (hunk 4 above)
Comment (ryanslade) [reply src/main/java/com/sourcegraph/webhook/WebhookListener.java:59]: Our current Go code mostly ignores those keys so I'll take this opportunity to clean it up. We don't actually care what kind of activity it is. 

We do still need the new key though so that we know when we've received a PullRequestParticipantStatusUpdatedEvent

Comment (ryanslade) [reply src/main/java/com/sourcegraph/webhook/WebhookListener.java:59]: Ah, we use those strings here too so I'll leave them there for now:
https://github.com/sourcegraph/bitbucket-server-plugin/blob/a312d669568da1aedf4903b9855ff2198b588bd2/src/main/java/com/sourcegraph/webhook/EventSerializer.java#L74

I wouldn't mind doing a proper refactor of this code but at this point I don't see it as a priority.
